core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,277 @@
+                    return false;
+                }
+                ComparisonExpression whenComparisonFunction = (ComparisonExpression) whenOperand;
+                Expression left = whenComparisonFunction.getLeft();
+                Expression right = whenComparisonFunction.getRight();
+
+                if (!whenComparisonFunction.getOperator().equals(EQUAL)) {
+                    return false;
+                }
+
+                if (commonLHSExpression == null) {
+                    commonLHSExpression = left;
+                }
+                else if (!commonLHSExpression.equals(left)) {
+                    return false;
+                }
+                rightHandSideExpressions.add(right);
+            }
+            return validateRHSExpressions(rightHandSideExpressions);
+        }
+
+        private boolean validateRHSExpressions(List<Expression> expressions)
+        {
+            Set<Expression> literals = new HashSet<>();
+            for (Expression expression : expressions) {
+                if (!isLiteral(expression) || literals.contains(expression)) {
+                    return false;
+                }
+                literals.add(expression);
+            }
+            return true;
+        }
+
+        private boolean isLiteral(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isLiteral(((Cast) expression).getExpression());
+            }
+            return expression instanceof Literal;
+        }
+
+        private Expression processCaseExpression(Optional<Cast> castExpression,
+                                                 List<WhenClause> whenClauses,
+                                                 Optional<Expression> defaultValue,
+                                                 Function<WhenClause, Expression> operandExtractor,
+                                                 Expression otherExpression,
+                                                 ComparisonExpression.Operator operator)
+        {
+            List<Expression> andExpressions = new ArrayList<>();
+            List<Expression> invertedOperands = new ArrayList<>();
+
+            for (WhenClause whenClause : whenClauses) {
+                Expression whenOperand = operandExtractor.apply(whenClause);
+                Expression whenResult = getCastExpression(castExpression, whenClause.getResult());
+
+                andExpressions.add(ExpressionUtils.and(
+                        new ComparisonExpression(operator, whenResult, otherExpression),
+                        whenOperand));
+                invertedOperands.add(new NotExpression(whenOperand));
+            }
+
+            Expression elseResult = defaultValue
+                    .map(x -> (Expression) new ComparisonExpression(operator, getCastExpression(castExpression, x), otherExpression))
+                    .orElse(new IsNullPredicate(otherExpression));
```

Comment:
This is not correct. We cannot replace any operator with "is null", when one operand is null.
It should rather be:
```
Expression default = defaultValue
        .map(value -> getCastExpression(castExpression, value))
        .orElse(<cast null to the expected type>);

Expression elseResult = new ComparisonExpression(operator, default, otherExpression);

Reply:
Made it to return NullLiteral in case of missing else clause since figuring out the expected type from Optimizer is difficult and the Simplify rules applied later adds a CAST if necessary. Now the results are similar to how trino handles other NULL expressions involved comparison function (i.e, NULL > 'string')

I have disabled the unit test with respect to this change - testRewriterOnCaseExpressionWithoutElseClause.
The reason is, the test was failing because of the difference in Cast object created.

When the query is parsed by PlanBuilder.expression, typeOnly is set to false.
```
public Cast(NodeLocation location, Expression expression, DataType type, boolean safe)
{
    this(Optional.of(location), expression, type, safe, false);
}
```
But when the expression is optimized using ExpressionInterpreter, typeOnly is set to true.
```
// Code block from LiteralEncoder.toExpression
if (object == null) {
    if (type.equals(UNKNOWN)) {
        return new NullLiteral();
    }
    return new Cast(new NullLiteral(), toSqlType(type), false, true);
}
```
Not sure if this is an existing bug.

Reply:
> Not sure if this is an existing bug.

Setting `typeOnly` to false, and not revisiting it later for the expected expression, should be considered a simplification of the testing process rather than a bug. I agree, however, that it might get in the way. This issue was already encountered and addressed here: https://github.com/trinodb/trino/blob/7b0990fcd70b1fe2995caa1071916571b75693eb/core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestSimplifyExpressions.java#L390-L396

Reply:
Done.